packages/app/src/composables/useCloudSpecData.ts

```diff
@@ -0,0 +1,119 @@
+/**
+ * Set up watchers & subscriptions to clear/refetch cloud data based on project state relative to latest data in the cloud
+ * @param isProjectDisconnected Whether project has a valid connection to the cloud
+ * @param isOffline  Whether app is currently offline (network state)
+ * @param projectId Cloud project ID (slug)
+ * @param mostRecentUpdate Date/time of the latest project update in the cloud
+ * @param displayedSpecs Spec entries currently displayed in the app
+ * @param allSpecs All project spec entries
+ * @returns Trigger functions
+ */
+export function useCloudSpecData (
+  isProjectDisconnected: Ref<boolean>,
+  isOffline: Ref<boolean>,
+  projectId: string | null | undefined,
+  mostRecentUpdate: Ref<string | null>,
+  displayedSpecs: Ref<(SpecsListFragment | undefined)[]>,
+  allSpecs: (SpecsListFragment | undefined)[],
+) {
+  const refetchMutation = useMutation(CloudData_RefetchDocument)
+
+  const isCloudSpecOlderThan = (item: NonNullCloudSpec, comparisonDttm: string | null) => {
+    if (item.data?.__typename !== 'CloudProjectSpec' && item.data?.__typename !== 'CloudProjectSpecNotFound') {
+      return false
+    }
+
+    if (!item.data.retrievedAt || !comparisonDttm) {
+      return false
+    }
+
+    return new Date(comparisonDttm).getTime() > new Date(item.data.retrievedAt).getTime()
+  }
+
+  const shouldRefetch = (item: NonNullCloudSpec) => {
+    if (isOffline.value) {
+      // Offline, no need to refetch
+      return false
+    }
+
+    if (item.fetchingStatus === 'NOT_FETCHED' || item.fetchingStatus === undefined) {
+      // Not yet fetched
+      return true
+    }
+
+    if (isCloudSpecOlderThan(item, mostRecentUpdate.value)) {
+      // outdated
+      return true
+    }
+
+    // nothing new, no need to refetch
+    return false
+  }
+
+  /**
+   * Refetch any displayed RemoteFetchable entries that are older than the latest cloudProject update
+   * or that have not yet been fetched
+   */
+  const fetchDisplayedCloudData = async () => {
+    const cloudSpecIdsToRefetch = displayedSpecs.value
+    .map((spec) => spec?.cloudSpec)
+    .filter((cloudSpec): cloudSpec is NonNullCloudSpec => Boolean(cloudSpec && shouldRefetch(cloudSpec)))
+    .map((cloudSpec) => cloudSpec.id)
+    ?? []
+
+    if (!isProjectDisconnected.value && !refetchMutation.fetching.value && cloudSpecIdsToRefetch.length > 0) {
+      await refetchMutation.executeMutation({ ids: cloudSpecIdsToRefetch })
+    }
+  }
+
+  /**
+   * Refetch any cloudSpec entries that are in an Error state
+   */
+  const refetchFailedCloudData = async () => {
+    const latestRunsIds = allSpecs
+    .map((s) => s?.cloudSpec)
+    .filter((cloudSpec): cloudSpec is NonNullCloudSpec => Boolean(cloudSpec?.fetchingStatus === 'ERRORED'))
+    .map((cloudSpec) => cloudSpec.id) ?? []
+
+    await refetchMutation.executeMutation({ ids: [...latestRunsIds] })
+  }
+
+  const displayedSpecIds = computed(() => displayedSpecs.value.map((v) => v?.cloudSpec?.id).join('|'))
```

Comment:
What is `v?.cloudSpec` is null? Do we need to filter out those potentially? I guess since we just use this for a kind of caching comparison, it's not a big deal?

Reply:
Good callout - nil values wouldn't break anything in this case but could trigger unnecessary refetch calls. Will filter to truthy cloudSpec ids